rust/otap-dataflow/crates/otap/src/retry_processor.rs

```diff
@@ -323,208 +234,212 @@ pub fn create_retry_processor(
-        let id = rstate.id;
-        if let Some(_removed) = self.pending_messages.remove(&id) {
-            log::debug!("Acknowledged and removed message with ID: {id}");
-            if let Some(m) = self.metrics.as_mut() {
-                m.msgs_acked.inc();
-            }
-        } else {
-            log::warn!("Attempted to acknowledge non-existent message with ID: {id}");
-        }
-        Ok(())
+        Self { config, metrics }
     }
 
-    async fn handle_nack(
+    async fn handle_ack(
         &mut self,
-        nack: NackMsg<OtapPdata>,
-        _effect_handler: &mut EffectHandler<OtapPdata>,
+        ack: AckMsg<OtapPdata>,
+        effect_handler: &mut EffectHandler<OtapPdata>,
     ) -> Result<(), Error> {
-        let rstate: RetryState = nack.calldata.try_into()?;
+        let signal = ack.accepted.signal_type();
+        let calldata = ack.calldata.clone();
 
-        let id = rstate.id;
-        if let Some(m) = self.metrics.as_mut() {
-            m.nacks_received.inc();
-        }
-        if let Some(mut pending) = self.pending_messages.remove(&id) {
-            pending.retry_count += 1;
-            pending.last_error = nack.reason;
-
-            if pending.retry_count <= self.config.max_retries {
-                let delay_ms = (self.config.initial_retry_delay_ms as f64
-                    * self
-                        .config
-                        .backoff_multiplier
-                        .powi(pending.retry_count as i32 - 1))
-                .min(self.config.max_retry_delay_ms as f64) as u64;
-
-                pending.next_retry_time = Instant::now() + Duration::from_millis(delay_ms);
-                let retry_count = pending.retry_count;
-                let _previous = self.pending_messages.insert(id, pending);
-                if let Some(m) = self.metrics.as_mut() {
-                    m.retry_attempts.inc();
-                }
-                log::debug!("Scheduled message {id} for retry attempt {retry_count}");
-            } else {
-                log::error!(
-                    "Message {} exceeded max retries ({}), dropping. Last error: {}",
-                    id,
-                    self.config.max_retries,
-                    pending.last_error
-                );
-                if let Some(m) = self.metrics.as_mut() {
-                    m.msgs_dropped_exceeded_retries.inc();
-                }
+        effect_handler.notify_ack(ack).await?;
+
+        // Forward the ACK upstream to complete the request chain
+        match calldata.try_into() {
+            Err(_err) => {
+                // Malformed context error: we don't know what this is.
             }
-        } else {
-            log::warn!("Attempted to handle nack for non-existent message with ID: {id}");
-        }
+            Ok(RetryState { num_items, .. }) => {
+                // The producer and consumer both succeed.
+                self.metrics.add_consumed_success(signal, num_items);
```

Comment:
Shouldn't this metric be updated in the `process` method when you receive `Message::PData`?

Reply:
But if `produced_success` is always going to be equal to `consumed_success`, then maybe we just need one metric here. I don't see any additional value being added by tracking a duplicate metric in the form of `consumed_sucess`.

Reply:
Done! Thank you, that makes sense.

Another way I was thinking about this, is that if this were rewritten with "success", "failure", and "refused" as attributes instead of part in the metric name, then it would look a little more ordinary to have two metrics incremented with the same attribute values in one location. Either form, the attribute form or the metrics-looking-redundant form, is meant to facilitate a straightforward query experience where to define the success-rate of a consumer (ignoring retries) you use the consumer-side metrics (e.g., consumer{outcome=success}/consumer{outcome=.}). To evaluate the producer (considering retries), you use the producer-side metrics (e.g., producer{outcome=success}/producer{outcome=.}).

Reply:
#1291 is related.